oximeter/db/src/client.rs

```diff
@@ -384,30 +380,47 @@ impl Client {
         &self,
         sample: &Sample,
     ) -> Result<Option<String>, Error> {
-        let schema = model::schema_for(sample);
-        let name = schema.timeseries_name.clone();
-        let maybe_new_schema = match self.schema.lock().unwrap().entry(name) {
-            Entry::Vacant(entry) => Ok(Some(entry.insert(schema).clone())),
+        let sample_schema = model::schema_for(sample);
+        let name = sample_schema.timeseries_name.clone();
+        let mut schema = self.schema.lock().await;
+
+        // We need to possibly check that this schema is in the local cache, or
+        // in the database, all while we hold the lock to ensure there's no
+        // concurrent additions. This containment check is needed so that we
+        // check both the local cache and the database, to avoid adding a schema
+        // a second time.
+        if !schema.contains_key(&name) {
+            self.get_schema_locked(&mut schema).await?;
+        }
```

Comment:
Why is checking `schema.contains_key(&name)` sufficient to decide whether we need to go update the cached value from the DB? Can there not be a situation where the name is already in the cache but the associated schema in the DB is different? Or is that guarantee what locking buys us?

Reply:
I think that first paragraph is a much better comment than what's there!

Reply:
Or both!

Reply:
Sweet, will update

Reply:
Done in c1b287520